src/test/java/org/elasticsearch/index/shard/SimpleIndexShardCounterTests.java

```diff
@@ -0,0 +1,209 @@
+            }
+        });
+        sleep(50);
+        assertThat(indexShard.getRefCount(), equalTo(2));
+        indexResponseFuture.get();
+        assertTrue(awaitBusy(new Predicate<Object>() {
+            @Override
+            public boolean apply(Object o) {
+                return indexShard.getRefCount() == 1;
+            }
+        }));
+    }
+
+    @Test
+    public void testRemoveShardWaitsForEngineClose() throws InterruptedException, ExecutionException, IOException {
+        assertAcked(client().admin().indices().prepareCreate("test"));
+        ensureGreen("test");
+        IndicesService indicesService = getInstanceFromNode(IndicesService.class);
+        IndexService indexService = indicesService.indexServiceSafe("test");
+        IndexShard indexShard = indexService.shard(0);
+        logger.info("removing shard");
+        indexService.removeShard(indexShard.shardId().id(), "removing to check if counter dec");
+        assertTrue(indexShard.engineClosed());
+        assertThat(indexShard.getRefCount(), equalTo(0));
+    }
+
+    @Test
+    @TestLogging("org.elasticsearch.action:TRACE")
+    public void testRemoveShardCanCauseIndexMissingException() throws InterruptedException, ExecutionException, IOException {
+        assertAcked(client().admin().indices().prepareCreate("test").setSettings(ImmutableSettings.builder().put("index.number_of_shards", 1).put("index.number_of_replicas", 0))
+                .addMapping("doc", jsonBuilder()
+                        .startObject()
+                        .startObject("doc")
+                        .startObject("transform")
+                        .field("script", "sleep(200)")
+                        .field("lang", "groovy")
+                        .endObject()
+                        .endObject()
+                        .endObject())
+                .get());
+        ensureGreen("test");
+        IndicesService indicesService = getInstanceFromNode(IndicesService.class);
+        IndexService indexService = indicesService.indexServiceSafe("test");
+        IndexShard indexShard = indexService.shard(0);
+        ActionFuture<IndexResponse> indexResponseActionFuture = client().prepareIndex().setIndex("test").setType("doc").setSource("{\"text\":\"text\"}").execute();
+        sleep(50);
+        logger.info("removing shard");
+        indexService.removeShard(indexShard.shardId().id(), "removing to check if counter dec");
+        assertTrue(indexShard.engineClosed());
+        assertThat(indexShard.getRefCount(), equalTo(0));
+        try {
+            indexResponseActionFuture.get();
+            fail();
+        } catch (ExecutionException e) {
+            logger.info("caught a", e);
+            if (e.getCause() instanceof IndexMissingException == false) {
+                fail();
+            }
+        }
+    }
+
+    @Test
+    public void testCloseShardCanCauseIndexShardClosedException() throws InterruptedException, ExecutionException, IOException {
```

Comment:
these are all integration tests mostly with all the groovy stuff - I think we should not do this at all. I really don't like the sleeps here